tuf/tuf.go

```diff
@@ -296,114 +302,102 @@ func (tr *Repo) GetAllLoadedRoles() []*data.Role {
 	return res
 }
 
-// GetDelegation finds the role entry representing the provided
-// role name along with its associated public keys, or ErrInvalidRole
-func (tr *Repo) GetDelegation(role string) (*data.Role, data.Keys, error) {
-	if !data.IsDelegation(role) {
-		return nil, nil, data.ErrInvalidRole{Role: role, Reason: "not a valid delegated role"}
-	}
-
-	parent := path.Dir(role)
-
-	// check the parent role
-	if _, err := tr.GetDelegationRole(parent); parent != data.CanonicalTargetsRole && err != nil {
-		return nil, nil, data.ErrInvalidRole{Role: role, Reason: "parent role not found"}
-	}
-
-	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
-	if !ok { // the parent targetfile may not exist yet, so it can't be in the list
-		return nil, nil, data.ErrNoSuchRole{Role: role}
-	}
-
-	foundAt := utils.FindRoleIndex(p.Signed.Delegations.Roles, role)
-	if foundAt < 0 {
-		return nil, nil, data.ErrNoSuchRole{Role: role}
-	}
-	delegationRole := p.Signed.Delegations.Roles[foundAt]
-	keys := make(data.Keys)
-	for _, keyID := range delegationRole.KeyIDs {
-		keys[keyID] = p.Signed.Delegations.Keys[keyID]
-	}
-	return delegationRole, keys, nil
-}
-
 // UpdateDelegations updates the appropriate delegations, either adding
 // a new delegation or updating an existing one. If keys are
 // provided, the IDs will be added to the role (if they do not exist
 // there already), and the keys will be added to the targets file.
-func (tr *Repo) UpdateDelegations(role *data.Role, keys []data.PublicKey) error {
-	if !data.IsDelegation(role.Name) {
-		return data.ErrInvalidRole{Role: role.Name, Reason: "not a valid delegated role"}
+func (tr *Repo) UpdateDelegations(roleName string, changes changelist.TufDelegation) error {
+	if !data.IsDelegation(roleName) {
+		return data.ErrInvalidRole{Role: roleName, Reason: "not a valid delegated role"}
 	}
-	parent := path.Dir(role.Name)
+	parent := path.Dir(roleName)
 
 	if err := tr.VerifyCanSign(parent); err != nil {
 		return err
 	}
 
 	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
+	_, ok := tr.Targets[parent]
 	if !ok { // the parent targetfile may not exist yet - if not, then create it
 		var err error
-		p, err = tr.InitTargets(parent)
+		_, err = tr.InitTargets(parent)
```

Comment:
If we call `InitTargets` here, the created targets metadata will be marked as dirty.  If later, as we walk the tree, we figure out that the delegation is invalid (not within the target/delegations role hierarchy), this metadata may still be published.  Currently, walk continues if the metadata doesn't exist, but I'm wondering if it should just pass the nil pointer to the walk functions, so they can create a new metadata object and assign it if necessary?  It would prevent us from having to do a create before we actually walk the tree.

(Alternately, we can set `p.Dirty = false` here)